### PR TITLE
spec updates

### PR DESCRIPTION
**Pull request by arnetheduck**
random stuff encountered while reading spec

### CODE REVIEW

File: beacon_chain/private/helpers.nim [hunk 1]
```diff
@@ -8,102 +8,75 @@
...
+    index = 0
+  while index < values_count - 1:
+    # Re-hash the `source` to obtain a new pattern of bytes.
 
-  var i = 0
-  while i < values.len - 1:
-    # Re-hash the `source` to obtain a new pattern of bytes
-    source = blake2_256.digest source.data
-    # Iterate through the `source` bytes in 3-byte chunks
+    # XXX "attempting to call undeclared routine init"
```
Comment (mratsim) [inline beacon_chain/private/helpers.nim]: You probably need a `mixin init`

Comment (mratsim) [reply beacon_chain/private/helpers.nim]: Cf: https://github.com/cheatfate/nimcrypto/blob/master/nimcrypto/hash.nim#L60-L80 I think Nim tries to resolve the symbol too early

File: beacon_chain/private/helpers.nim [hunk 2]
```diff
@@ -8,102 +8,75 @@
...
-      committees[j].shard_id = (shard_id_start + j.uint16) mod SHARD_COUNT
-      committees[j].committee = indices
-
-    result.add committees
-
 func get_shards_and_committees_for_slot*(state: BeaconState,
-        slot: uint64): seq[ShardAndCommittee] =
+                                         slot: uint64
+                                         ): seq[ShardAndCommittee] =
   # TODO: Spec why is active_state an argument?
```
Comment (mratsim) [inline beacon_chain/private/helpers.nim]: this comment can be removed

File: beacon_chain/private/helpers.nim [hunk 3]
```diff
@@ -8,102 +8,75 @@
...
   # TODO: this returns a scalar, not vector, but its return type in spec is a seq/list?
```
Comment (mratsim) [inline beacon_chain/private/helpers.nim]: Seems to be fixed with this PR

File: beacon_chain/private/helpers.nim [hunk 4]
```diff
@@ -114,15 +87,19 @@ func get_block_hash*(state: BeaconState, current_block: BeaconBlock, slot: int):
...
   let d = current_slot - parent_slot
   result = old_block_hashes[d .. ^1]
   for _ in 0 ..< min(d, old_block_hashes.len):
     result.add parent_hash
 
+func get_beacon_proposer*(state: BeaconState, slot: uint64): ValidatorRecord =
+  let
+    first_committee = get_shards_and_committees_for_slot(state, slot)[0].committee
+    index = first_committee[(slot mod len(first_committee).uint64).int]
+  state.validators[index]
```
Comment (mratsim) [inline beacon_chain/private/helpers.nim]: I have implemented this in my own branch with several comments that I feel are helpful to explain what is going on:

https://github.com/status-im/nim-beacon-chain/blob/19571482f105d0d905b6a943798f6a40aa4c1026/beacon_chain/private/helpers.nim#L129-L140

```Nim
func get_beacon_proposer*(state: BeaconState, slot: uint64): ValidatorRecord =
  ## From Casper RPJ mini-spec:
  ## When slot i begins, validator Vidx is expected
  ## to create ("propose") a block, which contains a pointer to some parent block
  ## that they perceive as the "head of the chain",
  ## and includes all of the **attestations** that they know about
  ## that have not yet been included into that chain.
  ##
  ## idx in Vidx == p(i mod N), pi being a random permutation of validators indices (i.e. a committee)
  let first_committee = get_shards_and_committees_for_slot(state, slot)[0]
  let index = first_committee.committee[slot.int mod first_committee.committee.len]
  result = state.validators[index]
```

File: tests/test_validator.nim [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  ## For now just test that we can compile and execute block processing with mock data.
+
+  test "Smoke validator shuffling":
+    let
+      validators = repeat(
+        ValidatorRecord(
+          status: ACTIVE
+        ), 1024)
+
+    # XXX the shuffling looks really odd, probably buggy
```
Comment (mratsim) [inline tests/test_validator.nim]: Repurposed https://github.com/status-im/nim-beacon-chain/issues/1#issuecomment-441662920 for shuffling test implementations using Sigp tests as reference - https://github.com/sigp/lighthouse/blob/ba548e49a52687a655c61b443b6835d79c6d4236/beacon_chain/validator_shuffling/src/shuffle.rs

File: beacon_chain/private/helpers.nim (hunk 1 above)
Comment (arnetheduck) [reply beacon_chain/private/helpers.nim]: `mixin init` here doesn't seem to help - added a comment for now, will dig into it later

File: beacon_chain/private/helpers.nim (hunk 4 above)
Comment (arnetheduck) [reply beacon_chain/private/helpers.nim]: nice! added.

File: tests/test_validator.nim (hunk 5 above)
Comment (arnetheduck) [reply tests/test_validator.nim]: thanks! added links to code, to keep track

**Pull request merged**